Grain damage formulation

gassmoeller opened the pull request:
Adding another option to compute grain size evolution following Mulyukova & Bercovici 2018. Almost ready for a review by @jdannberg, but I will need to finalize the benchmark results first.

jdannberg commented on tests/grain_size_strain_pinned_state_2/screen-output:
```diff
@@ -0,0 +1,868 @@
+   Solving grain_size system ... 0 iterations.
+   Rebuilding Stokes preconditioner...
+   Solving Stokes system... 1+0 iterations.
+      Relative nonlinear residuals (temperature, compositional fields, Stokes system): 1.47844e-15, 4.31876e-13, 0.813543
```
I guess the Stokes residual still does not decrease. That's still something we need to figure out.

jdannberg commented on benchmarks/grain_size_plunge/README.md:
```diff
@@ -0,0 +1,12 @@
+It models the grain size evolution under constant stress in a passive margin.
+No analytical expression for the the time-evolution of the grain size exists, but
+plotting the evolution using the 'plot.plt' script produces a figure very close to
+Fig. 4.
```
Did you want to add the notebook that is relevant to this benchmark as well? I think that would be useful.

jdannberg commented on benchmarks/grain_size_plunge/grain_size_plunge.prm:
```diff
@@ -0,0 +1,171 @@
+# from Mulyukova and Bercovici (2018) at constant
+# temperature and stress in their figure 4.
+
+#include grain_size_strain.prm
```
Just noticed this: If that's not needed, remove that line.

jdannberg commented:
Try this:

`set Nonlinear solver scheme = iterated Advection and defect correction Stokes`

That seems to help with the nonlinear convergence (no idea what the problem with the original residual computation is...)

gassmoeller commented on contrib/grain_size/equilibrium_grain_size_pinned_state.ipynb:
```diff
@@ -0,0 +1,181 @@
+    "shear_heating = 1.3194e-08\n",
+    "grain_reduction1 = 1.16823e-14\n",
+    "grain_growth1 = 9.24506e-24\n",
+    "equilibrium_grain_size1 = (grain_growth1/grain_reduction1)**(1/5)"
```
I dont remember exactly, but I think it was debugging output from ASPECT internal variables. We stored this to compare the size of the individual grain size change terms to the theoretical predictions, and to check if our ASPECT equilibrium grain size is consistent with the equations we derived in this notebook. I deleted the whole part, because it is impossible for the user to recreate and the dynamic test fulfils the same purpose.

gassmoeller commented:
I addressed all comments and updated the tests and figures. I moved all the .ipynb files into the benchmarks folder and wrote a readme that explains the relation between the tests, the benchmark, and the python notebooks. This should be ready for another round of review.

jdannberg commented on benchmarks/grain_size_pinned_state/dynamic_grain_size_pinned_state.ipynb:
```diff
@@ -0,0 +1,163 @@
+        "This notebook reports on the test results for test `grain_size_strain_pinned.prm` and allows to compare the test against analytical results. The test models a small volume in simple shear flow with constant strain rate, and a grain size that is not at its equilibrium value. We can compute how much the grain size should change according to the terms given in Eq. 8 of Mulyukova and Bercovi ...
+        "\n",
+        "\\begin{align}\n",
+        "\\text{grain\\_growth} = \\frac {3 \\phi_1\\phi_2 k_g \\sqrt h_g^{-q}}{(q R^{q-1})} \\exp(\\frac{-E_g}{R_gT}) \\\\\n",
```
I do not think you need the double \ before the underscore. At least it now has lots of `\_` in the notebook.

jdannberg commented on benchmarks/grain_size_pinned_state/equilibrium_grain_size_pinned_state.ipynb:
```diff
@@ -0,0 +1,178 @@
+    "Combining the prefactors of R (grain_size) into one such that the above equation can be written as :\n",
+    "\n",
+    "\\begin{equation}\n",
+    "0 = \\frac{\\text{grain\\_growth\\_term}}{R^{(q-1)}} - \\text{grain\\_reduction\\_term}*R^2\n",
```
Same in this notebook, I think you need to remove one \ before the underscores.

jdannberg commented on benchmarks/grain_size_pinned_state/phase_distribution_function.ipynb:
```diff
@@ -0,0 +1,143 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(\"Computed conversion factor: \", mean_grain_size_to_roughness_factor(0.6))\n",
```
Does this actually run? You renamed the function definition, but I think this is still the old name.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -1211,6 +1334,35 @@ namespace aspect
                              "to compute material properties (slower but more accurate).");
+          prm.enter_subsection("Grain damage partitioning");
+          {
+            prm.declare_entry ("Minimum partitioning temperature", "1600",
+                               Patterns::Double (0.),
+                               "This parameter determines the temperature at which the computed coefficient of shear energy "
+                               "partitioned into grain damage is minimum. This is used in the pinned state limit of the grain "
+                               "size evolution. One choice of this parameter is the mantle temperature at the ridge axis, "
+                               "see Mulyukova and Bercovici (2018) for details.");
+            prm.declare_entry ("Maximum partitioning temperature", "283",
```
After seeing your Assert below, I now realize these names are not intuitive (i.e., a user would expect the minimum temperature to be smaller than the maximum temperature). 
Do you think it makes sense to come up with better names?

jdannberg reviewed (commented):
I still found a few things that I think should be fixed, but otherwise this looks good to me! Thank you for taking the time to clean up all of this!

jdannberg commented on benchmarks/grain_size_pinned_state/grain_size_plunge.prm:
```diff
@@ -0,0 +1,172 @@
+set Maximum first time step                = 10
+set Maximum relative increase in time step = 1
+set Nonlinear solver scheme                = iterated Advection and defect correction Stokes
+set Max nonlinear iterations = 1
```
Wait, why is this one?

gassmoeller commented:
I have addressed your comments. Here are replies:

> I do not think you need the double \ before the underscore. At least it now has lots of \_ in the notebook.

I was confused by your comment, because I need the double underscores, but it turns out that jupyter notebooks (e.g. on Google Colab) use Mathjax to render equations, and VS code uses KaTex, and they differ in their treatment of underscores in `\text{}` commands (see https://github.com/mathjax/MathJax/issues/1770). I rewrote the equations to avoid the use of underscores in text.

> Since your PR includes commits from others, I think it would be good to mention all contributors here.

Yes, I just took over a changelog entry when rebasing without realizing the entry was only for part of what all of us did.

> After seeing your Assert below, I now realize these names are not intuitive (i.e., a user would expect the minimum temperature to be smaller than the maximum temperature). Do you think it makes sense to come up with better names?

Do you have suggestions for better names? I tried to improve them a bit, but it is inherently hard, because the maximum temperature controls the minimum partition and vice versa.

>  Wait, why is this one?

Leftover from testing, good find.

jdannberg reviewed (approved):
Looks good to me!
